rq/queue.py

```diff
@@ -280,33 +280,61 @@ def enqueue_job(self, job, pipeline=None, at_front=False):
             job.timeout = self.DEFAULT_TIMEOUT
         job.save(pipeline=pipe)
 
+        if self._async:
+            self.push_job_id(job.id, pipeline=pipe, at_front=at_front)
+
         if pipeline is None:
             pipe.execute()
 
-        if self._async:
-            self.push_job_id(job.id, at_front=at_front)
-
         return job
 
     def enqueue_dependents(self, job, pipeline=None):
-        """Enqueues all jobs in the given job's dependents set and clears it."""
-        # TODO: can probably be pipelined
+        """Enqueues all jobs in the given job's dependents set and clears it.
+
+        When called without a pipeline, this method uses WATCH/MULTI/EXEC.
+        If you pass a pipeline, only MULTI is called. The rest is up to the
+        caller.
+        """
         from .registry import DeferredJobRegistry
 
+        pipe = pipeline if pipeline is not None else self.connection._pipeline()
+        dependents_key = job.dependents_key
+
         while True:
-            job_id = as_text(self.connection.spop(job.dependents_key))
-            if job_id is None:
+            try:
+                # if a pipeline is passed, the caller is responsible for calling WATCH
+                # to ensure all jobs are enqueued
+                if pipeline is None:
```

Comment:
I don't understand it completely:
The optional `pipe.watch()` is part of the code that allows to call `enqueue_dependents` standalone. Ok that's some kind of nice to have which could be kicked, since it is only used by the worker, which needs the pipeline to be race condition free.
But even if we kick the possibility of calling it standalone, I think the `watch` should be called by the worker to understand, where the `WatchError` can come from. But ok, that could be documented?
Reverting the code in `Worker.perform_job` adds the race condition again (enqueuing of the dependants AND the status change has to happen in one transaction). So where should I put it?